Make Map subscriptable

antonagestam opened the pull request:
Fixes #12

1st1 commented on immutables/_map.c:
```diff
@@ -3363,6 +3363,11 @@ map_reduce(MapObject *self)
+static PyObject *
+map_py_class_getitem(PyObject *type, PyObject *item)
+{
+    return type;
```
There's a ref leak here, add Py_INCREF

1st1 replied:
Or better yet just use Py_RETURN_NONE macro

antonagestam replied:
I tried following [advice I got from](https://gitter.im/python/typing?at=5d25bd59cdb70805c96e239f) Ivan in the mypy team:

> (also it should return class object itself, not None)

Would this be the correct usage of `Py_INCREF`?

1st1 replied:
Yes.

antonagestam replied:
Also are you fine with that implementation as opposed to returning None? :)

1st1 replied:
Yes! :)

1st1 commented on immutables/map.py:
```diff
@@ -431,7 +432,12 @@ def __iter__(self):
+D = TypeVar('D', bound=Any)
+
+
+class Map(Mapping[K, V]):
```
Do we need this if we add a pyi file anyways?

antonagestam replied:
It's not entirely clear to me in what circumstances this implementation is used as opposed to the C one, but if this ever is used it should probably inherit from Mapping so that `Map[foo, bar]` doesn't raise a TypeError. Note that I am not experienced with writing C extensions and so this was mostly a hack from my point of view. If you feel like things look weird, they probably are but I am happy for the feedback and can put some time into getting it right :)

1st1 replied:
We can't inherit it from Mapping, as we can only do that for the pure Python version.  So both shouldn't be inherited from Mapping.

I'm working on a PEP to have `immutables.Map` in the stdlib.  If we do that we can just call `Mapping.register(Map)` and be done with this.

antonagestam replied:
Oh right, so should we make the pure Python version also implement `__class_get_item__` instead? Because I believe `Mapping.register(Map)` is not sufficient for making `Map[foo, bar]` return `Map`?

Very cool with a PEP and stdlib support! :)

1st1 commented:
> Oh right, so should we make the pure Python version also implement __class_get_item__ instead?

Yes, that's the idea.

> Because I believe Mapping.register(Map) is not sufficient for making Map[foo, bar] return Map?

True.  We only need the `Mapping.register()` part to make `isinstance(immumap, Mapping)` work.

antonagestam commented:
@1st1 Gotcha, thanks for taking the time to explain! I'll try to update the PR soon :)

antonagestam commented:
I added fixes that addresses your comments as well as a test case. Perhaps the documentation should be updated with an example as well?

1st1 commented on immutables/_map.c:
```diff
@@ -3375,6 +3381,7 @@ static PyMethodDef Map_methods[] = {
     {"update", (PyCFunction)map_py_update, METH_VARARGS | METH_KEYWORDS, NULL},
     {"__reduce__", (PyCFunction)map_reduce, METH_NOARGS, NULL},
     {"__dump__", (PyCFunction)map_py_dump, METH_NOARGS, NULL},
+    {"__class_getitem__", (PyCFunction)map_py_class_getitem, METH_O|METH_CLASS, NULL},
```
Please keep line length under 79 characters.

1st1 commented on immutables/_map.pyi:
```diff
@@ -0,0 +1,16 @@
+from typing import Mapping, TypeVar, Any, Hashable, Iterator, Union
+
+K = TypeVar('K', bound=Hashable)
```
Please add one extra empty line before `K =`

1st1 commented on immutables/_map.pyi:
```diff
@@ -0,0 +1,16 @@
+    def get(self, key: K, default: D = None) -> Union[V, D]: ...
+    def __getitem__(self, key: K) -> V: ...
+    def __contains__(self, key: object) -> bool: ...
+    def __iter__(self) -> Iterator[K]: ...
```
We also need to add `MapMutation` and `Map.mutate()` method.

1st1 commented:
Looks great! Only a few nits left.

antonagestam commented:
I have a few questions on the mutation API.

Should Map.mutate() always return a MapMutation with it's own signature? E.g. `Map[A, B]().mutate() => MapMutation[A, B]`? I think that would make sense as then MapMutation.finish() would return the same type Map.

As for both MapMutation and Map should I add type hints for all methods or just the ones that are part of the public API? There's a few methods that always raise an exception, like MapMutation.__iter__(), should these have the return type `typing.NoReturn` or just not be hinted?

I'm also curious as to why MapMutation.pop() accepts a variadic *args and then raises TypeError if it's longer than 1. Why not use a non-variadic single `default` argument? :)

1st1 commented:
> Should Map.mutate() always return a MapMutation with it's own signature? E.g. Map[A, B]().mutate() => MapMutation[A, B]?

Yes.

> As for both MapMutation and Map should I add type hints for all methods or just the ones that are part of the public API? 

Ideally, yes.

> There's a few methods that always raise an exception, like MapMutation.__iter__(), should these have the return type typing.NoReturn or just not be hinted?

I guess using `typing.NoReturn` makes sense.

> I'm also curious as to why MapMutation.pop() accepts a variadic *args and then raises TypeError if it's longer than 1. Why not use a non-variadic single default argument? :)

This is to emulate a "positional-only argument", IIRC.

antonagestam commented:
I added more type hints, so the mutation API should now be covered as well as the keys, values and items methods. I added a `BitmapNode` as an empty class because it is exposed as the type of the `m` arguments in `__init__` methods. However, it's implementation was not to decipher and so I left it untyped for now. I don't think I'll be able to figure out the correct type hints for it.

asvetlov commented on immutables/_map.pyi:
```diff
@@ -0,0 +1,87 @@
+class BitmapNode: ...
+
+
+class MapKeys(Generic[K]):
```
I guess mapping views should derive from corresponding `collections.abc`.
```suggestion
class MapKeys(Generic[K], collections.abc.KeysView):
```
in this case

antonagestam replied:
There's a corresponding generic type that we could use: https://docs.python.org/3/library/typing.html#typing.KeysView

Looking at [this table in the collections docs](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) it looks like the views also need to implement `__contains__` to fulfill the interface?

1st1 commented:
@ambv

antonagestam commented:
Ping me if you need me to rebase or make other changes :)

1st1 commented:
Merged.

@antonagestam thank you for your hard work!

@ambv please test this with edgedb codebase.  If all good, I'll release a new version.

ambv commented:
@1st1, so far, so good. Tested with EdgeDB and Mypy, behaves well.

1st1 commented:
Alright, I'll schedule a release now.

1st1 commented:
> @1st1, so far, so good. Tested with EdgeDB and Mypy, behaves well.

immutables@0.10 is out, you can now bump the EdgeDB dep.

jab commented on immutables/_map.pyi:
```diff
@@ -0,0 +1,87 @@
+
+
+K = TypeVar('K', bound=Hashable)
+V = TypeVar('V', bound=Any)
+D = TypeVar('D', bound=Any)
```
Do the "`bound=Any`" here buy anything?

antonagestam replied:
Just a matter of style, I prefer to more clearly communicate whenever type vars are unbound.

jab commented on immutables/_map.pyi:
```diff
@@ -0,0 +1,87 @@
+
+
+K = TypeVar('K', bound=Hashable)
+V = TypeVar('V', bound=Any)
```
Should the `V` TypeVar set `covariant=True`? That would match https://github.com/python/typeshed/blob/master/stdlib/3/typing.pyi#L413 better.

antonagestam replied:
Yes, that seems reasonable.

antonagestam commented on immutables/_map.pyi:
```diff
@@ -0,0 +1,87 @@
+from typing import Union
+
+
+K = TypeVar('K', bound=Hashable)
```
I think my intention here was that having that bound will give the user a type error for non-hashable keys instead of a runtime error. Perhaps this
